src/category_theory/adjunction/fully_faithful.lean

```diff
@@ -63,11 +64,87 @@ instance counit_is_iso_of_R_fully_faithful [full R] [faithful R] : is_iso (adjun
     simp,
   end⟩⟩⟩
 
--- TODO also prove the converses?
--- def L_full_of_unit_is_iso [is_iso (adjunction.unit h)] : full L := sorry
--- def L_faithful_of_unit_is_iso [is_iso (adjunction.unit h)] : faithful L := sorry
--- def R_full_of_counit_is_iso [is_iso (adjunction.counit h)] : full R := sorry
--- def R_faithful_of_counit_is_iso [is_iso (adjunction.counit h)] : faithful R := sorry
+/-- If the unit of an adjunction is an isomorphism, then its inverse on the image of L is given
+by L whiskered with the counit. -/
+@[reducible]
+def whisker_left_L_counit_iso_of_is_iso_unit [is_iso (adjunction.unit h)] : (L ⋙ R ⋙ L) ≅ L :=
+⟨ ⟨(whisker_left L h.counit).app, (whisker_left L h.counit).naturality⟩,
+  ⟨(whisker_right h.unit L).app, (whisker_right h.unit L).naturality⟩,
+  begin
+    ext x, dsimp,
+    rw [comp_hom_eq_id, ←hom_comp_eq_id],
+    simpa [adjunction.right_triangle, functor.id_obj],
+  end, begin
+    ext x, dsimp,
+    simp [adjunction.right_triangle, functor.id_obj],
+  end⟩
+
+/-- If the counit of an adjunction is an isomorphism, then its inverse on the image of R is given
+by R whiskered with the unit. -/
+@[reducible]
+def whisker_left_R_unit_iso_of_is_iso_counit [is_iso (adjunction.counit h)] : R ≅ (R ⋙ L ⋙ R) :=
+⟨ ⟨(whisker_left R h.unit).app, (whisker_left R h.unit).naturality⟩,
+  ⟨(whisker_right h.counit R).app, (whisker_right h.counit R).naturality⟩,
+  begin
+      ext x, dsimp,
+      simp [adjunction.left_triangle, functor.id_obj],
+  end, begin
+      ext x, dsimp,
+      rw [hom_comp_eq_id, ←comp_hom_eq_id],
+      simpa [adjunction.right_triangle, functor.id_obj],
+  end⟩
+
+/-- If the unit is an isomorphism, then the left adjoint is full-/
+noncomputable
+def L_full_of_unit_is_iso [is_iso (adjunction.unit h)] : full L :=
+⟨ λ X Y f, ((h.hom_equiv X (L.obj Y)) f) ≫ (inv (adjunction.unit h)).app Y,
+  λ X Y f,
+  begin
+    simp, symmetry,
+    suffices H : f ≫ L.map (h.unit.app Y) ≫ ((whisker_left_L_counit_iso_of_is_iso_unit h).hom.app Y)
+      = L.map (h.unit.app X) ≫ L.map (R.map f)
+        ≫ ((whisker_left_L_counit_iso_of_is_iso_unit h).hom.app Y),
+    { rw [←assoc, ←assoc] at H,
+      rw [(nat_iso.cancel_nat_iso_hom_right (whisker_left_L_counit_iso_of_is_iso_unit h)
+      (f ≫ L.map (h.unit.app Y)) (L.map (h.unit.app X) ≫ L.map (R.map f)))] at H,
+      rw [←assoc, is_iso.eq_comp_inv],
+      exact H},
+    { simp }
+  end⟩
```

Comment:
```suggestion
noncomputable def L_full_of_unit_is_iso [is_iso (adjunction.unit h)] : full L :=
{ preimage := λ X Y f, h.hom_equiv X (L.obj Y) f ≫ inv (h.unit.app Y) }
```
Using the lemma (+dual) I mentioned above, this proof simplifies a lot! A couple of other notes: when constructing a structure which has named fields, it's usually clearer to use their names in structure notation rather than using angle brackets.

Reply:
How come the `witness'` field does not even have to be specified? Is it because of the `."obviously"` in the definition of this field in `category_theory.full`?

Reply:
Exactly! Using structure notation allows the `. obviously` (known as an auto-parameter) to be filled in automatically, (almost) always by the tactic `tidy`: approximately `tidy` is a tactic which just does intros/cases/ext/dsimp/simp until the goal is done, so `tidy` being able to solve a goal like this is a sign that we have the right simp lemmas.

Reply:
You'll see auto-params showing up loads in category theory, they're really convenient to solve "boring" goals like this one

Reply:
Thanks! That's good to know!